File: day1/src/day1.py


digits = {'one': '1', 'two': '2', 'three': '3', 'four': '4', 'five': '5', 'six': '6', 'seven': '7', 'eight': '8', 'nine': '9'}

def func():
    text_file = open('day1\\data\\test-input-2','r')
    line_list = text_file.readlines()
    count = 0
    #for each line from the list, print the line
    for line in line_list:
        #print("the line: " + line)
        line = convertStringToDigit(line)
        val = int(getDigits(line))
        
        count = count + val
        print(line.strip() + " " + str(val) + " " + str(count))
    text_file.close() 


def convertStringToDigit(line):
    
    idx_map = {}
    leftIdx = len(line) - 1
    rightIdx = -1
    leftStr = ''
    rightStr = ''
    
    for i in digits:
        idx = line.find(i)
        if (line.find(i) >= 0):
            #print("found: ", i)
            if (idx < leftIdx):
                leftIdx = idx
                leftStr = i

            
    len_to_skip = len(leftStr)
    line = line[:leftIdx] + digits[leftStr] + line[leftIdx + len_to_skip:]

    print("after left replace: ", line)
    for i in digits:
        
        idx = line.rfind(i)
        if (idx >= 0):
            #print("found: ", i)
            if (idx > rightIdx):
                rightIdx = idx
                rightStr = i
    
    #line = line.replace(i, digits[i])
    
    if (rightStr != ""):
        rightIdx = rightIdx - len_to_skip + 1
        len_to_skip = len(rightStr)
        line = line[:rightIdx] + digits[rightStr] + line[rightIdx + len_to_skip:]
        print("after right replace: ", line)

    return line        
            

def getDigits(line):
    
    f =''
    l = ''
    for i in line:
        if i.isdigit():
            f = i

    line = line[::-1]
    for i in line:
        if i.isdigit():
            l = i

    return l+f


func()

File: day1/src/day1.java
package day1.src;

import java.io.File;
import java.io.FileNotFoundException;
import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;

public class day1 {

    private static Map<String, String> digitsMap = new HashMap<>(); 
    public static void main(String[] args) {
        try {
            File myObj = new File("day1\\data\\input");
            Scanner myReader = new Scanner(myObj);
            int counter = 0;
            
            digitsMap.put("one", "1");
            digitsMap.put("two", "2");
            digitsMap.put("three", "3");
            digitsMap.put("four", "4");
            digitsMap.put("five", "5");
            digitsMap.put("six", "6");
            digitsMap.put("seven", "7");
            digitsMap.put("eight", "8");
            digitsMap.put("nine", "9");

            while (myReader.hasNextLine()) {
                String data = myReader.nextLine();
                int val = getDigits(data);
                System.out.print(data);
                System.out.println(" --->> 6val: " + val);
                counter += val;
            }
            System.out.println("final val: " + counter);
        myReader.close();
        } catch (FileNotFoundException e) {
            System.out.println("An error occurred.");
            e.printStackTrace();
        } 
    }


    private static String replaceDigits(String data) {
        
        for (String str : digitsMap.keySet())
        if (data.indexOf(str) >= 0) {
            data.replace(str, digitsMap.get(str));
            break;
        }

        if (data.lastIndexOf(str) >= 0) {
            data.replace(str, digitsMap.get(str));
            break;
        }

        return "";
    }

    private static int getDigits(String data) {
        
        char[] chars = data.toCharArray();

        char first = '\0';
        char second = '\0';

        for (int i = 0; i < data.length(); i++) {
            if (Character.isDigit(chars[i])) {
                first = chars[i];
                break;
            }
        }

        for (int j = data.length() - 1; j >= 0; j-- ) {
            if (Character.isDigit(chars[j])) {
                second = chars[j];
                break;
            }
        }
        String str = "" + first + second;
        return Integer.parseInt(str);
    }
}
